plugin/evm/atomic_trie.go

```diff
@@ -354,11 +349,13 @@ func (a *atomicTrie) AcceptTrie(height uint64, root common.Hash) (bool, error) {
 		hasCommitted = true
 	}
 
-	// Attempt to dereference roots at least [tipBufferSize] old
-	//
-	// Note: It is safe to dereference roots that have been committed to disk
-	// (they are no-ops).
-	a.tipBuffer.Insert(root)
+	// The following dereferences, if any, the previously inserted root.
+	// This one can be dereferenced whether it has been:
+	// - committed, in which case the dereference is a no-op
+	// - not committted, in which case the current root we are inserting contains
+	//   references all the relevant data from the previous root, so the previous
+	//   root can be dereferenced.
+	a.trieDB.Dereference(a.lastAcceptedRoot)
```

Comment:
Note `a.lastAcceptedRoot` is set at the bottom of this function.
In case of commit errors, it's unclear what should be done though (it was unclear before as well), but I could move the lastAcceptedRoot field update right after this line if necessary.

Reply:
should we move this below the last `a.commit` on line 362 so it cannot fail anymore before we dereference?